hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/ErrataMigration.java

```diff
@@ -0,0 +1,196 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.flywaydb.core.api.MigrationVersion;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcOperations;
+
+import com.hedera.mirror.common.domain.balance.AccountBalanceFile;
+import com.hedera.mirror.common.domain.transaction.RecordItem;
+import com.hedera.mirror.importer.MirrorProperties;
+import com.hedera.mirror.importer.exception.FileOperationException;
+import com.hedera.mirror.importer.parser.balance.BalanceStreamFileListener;
+import com.hedera.mirror.importer.parser.record.RecordStreamFileListener;
+import com.hedera.mirror.importer.parser.record.entity.EntityRecordItemListener;
+import com.hedera.mirror.importer.reader.ValidatedDataInputStream;
+import com.hedera.mirror.importer.repository.TransactionRepository;
+
+@Log4j2
+@Named
+@RequiredArgsConstructor(onConstructor_ = {@Lazy})
+public class ErrataMigration extends MirrorBaseJavaMigration implements BalanceStreamFileListener {
+
+    @Value("classpath:errata/mainnet/balance-offsets.txt")
+    private final Resource balanceOffsets;
+    private final EntityRecordItemListener entityRecordItemListener;
+    private final NamedParameterJdbcOperations jdbcOperations;
+    private final MirrorProperties mirrorProperties;
+    private final RecordStreamFileListener recordStreamFileListener;
+    private final TransactionRepository transactionRepository;
+    private final Set<Long> timestamps = new HashSet<>();
+
+    @Override
+    public Integer getChecksum() {
+        return 1; // Change this if this migration should be rerun
+    }
+
+    @Override
+    public String getDescription() {
+        return "Add errata information to the database to workaround older, incorrect data";
+    }
+
+    @Override
+    public MigrationVersion getVersion() {
+        return null; // Repeatable migration
+    }
+
+    @Override
+    public void onStart() {
+        // Not applicable
+    }
+
+    @Override
+    public void onEnd(AccountBalanceFile accountBalanceFile) {
+        if (isMainnet() && getTimestamps().contains(accountBalanceFile.getConsensusTimestamp())) {
+            accountBalanceFile.setTimeOffset(-1);
+        }
+    }
+
+    @Override
+    public void onError() {
+        // Not applicable
+    }
+
+    @Override
+    protected void doMigrate() throws IOException {
+        if (isMainnet()) {
+            balanceFileAdjustment();
+            missingTransactions();
+            spuriousTransfers();
+        }
+    }
+
+    private void balanceFileAdjustment() {
+        String sql = "update account_balance_file set time_offset = -1 where consensus_timestamp in (:timestamps)";
```

Comment:
```suggestion
        String sql = "update account_balance_file set time_offset = -1 where consensus_timestamp in (:timestamps) and time_offset <> -1";
```
nit: avoid uncessary updates?

Reply:
Added